### ISSUE TITLE
Issue in handling delta jobs which are being created at the time of full Indexing.

### ISSUE THREAD

**Issue by deepak-tiwari** [pat/ts-delayed-delta#49]
Scenario:
When full indexer runs it first clears the delta jobs from the DJ Queue.
After that full indexer starts re-indexing the sphinx data. And during this time if any updates happen then there will two job added in the DJ Queue for each record update.
1.ThinkingSphinx::Deltas::DelayedDelta::DeltaJob
2.ThinkingSphinx::Deltas::DelayedDelta::FlagAsDeletedJob

Once the full indexer completes it toggle the delta flag of the updated records to false. But there will be Delta Jobs in the DJ Queue.

Now for a particular updated record this is how delta jobs being handled.
1. When ThinkingSphinx::Deltas::DelayedDelta::DeltaJob executes which generally add the document in the delta core if delta is true. But for this case delta will be false since full indexer has run just before. So it won't add the document in delta core.
2. When ThinkingSphinx::Deltas::DelayedDelta::FlagAsDeletedJob runs it will mark the document in the main core to be deleted irrespective of delta flag.

Now if we try to search those updated records it won't be available in search.

So, is it possible to have delta flag check in the ThinkingSphinx::Deltas::DelayedDelta::FlagAsDeletedJob too to fix the issue.? 
or you have any other suggestion to fix the problem.

**Comment by pat** [comment]
I've just pushed a commit that checks whether an instance still has `delta` set to true before marking it as deleted. I hope this resolves your problem - feel free to give it a spin:

```ruby
gem 'ts-delayed-delta', '~> 2.0.2',
  :git    => 'git://github.com/pat/ts-delayed-delta.git',
  :branch => 'master',
  :ref    => '357ca9e0c0'
```

And thank you for the detailed debugging - greatly appreciated!

**Comment by deepak-tiwari** [comment]
Hi Pat,
Thank you very much for the assistance. 

Also, after some testing we found that there is one more case which should also be taken care of with the above fix.

Scenario: Object Deletion
In ThinkingSphinx::Deltas::DelayedDelta::FlagAsDeletedJob class
def perform
    return unless instance.delta?
    ThinkingSphinx::Deltas::DeleteJob.new(index, document_id).perform
 end
def instance
    instance_type.constantize.find instance_id
end

Problem here is calling instance will throw exception in destroy case and corresponding document will still be present in the main core.

To fix the issue we can have extra check for instance presence.
If instance is not present we can directly call DeleteJob for the document.

Thank you again!

**Comment by pat** [comment]
So, the following would do the job?

```ruby
def perform
  return unless instance.nil? || instance.delta?

  ThinkingSphinx::Deltas::DeleteJob.new(index, document_id).perform
end

def instance
  instance_type.constantize.find instance_id
rescue ActiveRecord::RecordNotFound
  nil
end
```

**Comment by deepak-tiwari** [comment]
Yes. That's right Pat.

**Comment by pat** [comment]
Just pushed d1835499773274921e9672b3dab5d33a3da03c40 with that fix :)

**Comment by deepak-tiwari** [comment]
Not important but i think we can avoid calling db query on instance twice :-)
 ```ruby
def instance
     @instance || = instance_type.constantize.find_by_id instance_id
end
```

**Comment by pat** [comment]
Good point, patched in 4a868f2b4b3dce8bee832cd9ec21692c1056660a.

I avoided using `find_by_id` because it's not present in Rails 4.2 (or at least it's deprecated, and will be removed by 5.0). This gem needs to work with a lot of Rails versions, hence will just catch the exception instead.

**Comment by deepak-tiwari** [comment]
Awesome. Thank you very much :)
